governor.py

```diff
@@ -0,0 +1,408 @@
+
+    def get_status(self, flush=False):
+        """
+        Return limiter statuses and flush limiters
+        """
+        try:
+            statuses = [l.get_status() for l in self._limiters]
+        except TypeError:
+            log.exception("Governor instantiated before being set.")
+            return []
+
+        # Flush limiters
+        if flush:
+            for l in self._limiters:
+                l.flush(self._iteration - self._CONTEXTS_TTL)
+            self._iteration += 1
+
+        return statuses
+
+    def _name_args(self, arg_list, kwargs, asynchronous=False):
+        """
+        Name `arg_list` items and merge with `kwargs`
+        """
+        named_args = kwargs.copy()
+        for i, arg_value in enumerate(arg_list):
+            if asynchronous:
+                arg_name = self._METRIC_FORMAT[i]
+            else:
+                arg_name = self._submit_metric_arg_names[i + 1]
+            named_args[arg_name] = arg_value
+        return named_args
+
+    def _check(self, args):
+        """
+        Check metric against all limiters
+        """
+        try:
+            return all(r.check(self._iteration, args) for r in self._limiters)
+        except TypeError:
+            log.exception("Governor instantiated before being set.")
+            return True
+
+
+class CheckGovernor(Governor):
+    """
+    Monitor check's metric payload
+    """
+    def __init__(self):
+        super(CheckGovernor, self).__init__()
+        self._limiters = copy.deepcopy(self._CHECK_LIMITERS)
+
+    def process(self, metrics):
+        """
+        Asynchronous check's metric payload analysis. Do not report on anomalies.
+        """
+        return super(CheckGovernor, self).process(metrics, flush=False, report=False)
+
+    #####################################################
+    # Governor as a decorator  for `real-time` analysis #
+    #####################################################
+    # Deprecated: will be removed once confirmed #
+    ##############################################
+    def set(self, func):
+        """
+        Set governor to run as a decorator
+        """
+        self._submit_metric = func
+        self._submit_metric_arg_names = inspect.getargspec(func)[0]
+
+    def __call__(self, *args, **kw):
+        """
+        Decorator purpose around `submit_metric` method for 'real-time' analysis
+        """
+        # Shortcut when no rules are defined
+        if not self._limiters:
+            return self._submit_metric(*args, **kw)
+
+        # Really dirty trick -> to improve
+        named_args = self._name_args(args, kw)
+
+        # Extract argument dict
+        if self._check(named_args):
+            return self._submit_metric(*args, **kw)
+
+
+class AgentGovernor(Governor):
+    """
+    Monitor entire collector's metric payload
+    """
+    def __init__(self):
+        super(AgentGovernor, self).__init__()
+        self._limiters = copy.deepcopy(self._AGENT_LIMITERS)
```

Comment:
It will.  Limiters have a "state", in particular
* An index to keep track of active selections
* A trace, which records the number of metrics overflowing the limit set

As there should be only one `AgentGovernor` object instantiated, it would not be a problem to manipulate these class variables directly (unlike `CheckGovernor`). 
I still think it's preferable to keep them "blank" at class level, and make a copy for the instance.